src/merkle_tree.rs

```diff
@@ -1,25 +1,115 @@
-use sha3::{Keccak256, Digest};
+use sha3::{Digest, Sha3_256};
 
-struct Leaf {
+type Position = usize;
+
+#[derive(Debug)]
+pub struct Leaf {
     // The hash is an array of 32 u8
     hash: [u8; 32],
 
     // Each leaf will have an index to its children.
     // It's an option type because maybe the child does not
     // exist.
-    left_child: Option<usize>,
-    right_child: Option<usize>
+    left_child: Option<Position>,
+    right_child: Option<Position>,
+}
+
+impl Leaf {
+    pub fn new(
+        hash: [u8; 32],
+        left_child: Option<Position>,
+        right_child: Option<Position>,
+    ) -> Leaf {
+        Leaf {
+            hash,
+            left_child,
+            right_child,
+        }
+    }
+}
+
+pub fn hash(positions: [Position; 2], leaves: &[&Leaf; 2]) -> Leaf {
+    let new_hash = leaves
+        .iter()
+        .fold(Sha3_256::new(), |mut acc, a| {
+            let value = a.hash;
+            acc.update(value);
+            acc
+        })
+        .finalize();
+
+    Leaf {
+        hash: new_hash.into(),
+        right_child: Some(positions[1]),
+        left_child: Some(positions[0]),
+    }
 }
 
-struct MerkleTree {
+pub struct MerkleTree {
     leaves: Vec<Leaf>,
 }
 
-pub fn hello_world() {
-    let mut hasher = Keccak256::new();
-    let data = b"Hello world!";
-    println!("Binary hash: {:?}", data);
-    hasher.update(data);
-    let hash = hasher.finalize();
-    println!("Binary hash: {:?}", hash);
+impl MerkleTree {
+    pub fn new<T: std::convert::AsRef<[u8]>>(data: &[T]) -> MerkleTree {
+        let original_leaves: Vec<(Position, Leaf)> = data
+            .iter()
+            .map(|value| {
+                let hash = Sha3_256::digest(value);
+                hash
+            })
```

Comment:
this can be rewritten as `Sha3_256::digest(value)` if you are returning the hash